include/engine/routing_algorithms.hpp

```diff
@@ -0,0 +1,163 @@
+#ifndef OSRM_ENGINE_ROUTING_ALGORITHM_HPP
+#define OSRM_ENGINE_ROUTING_ALGORITHM_HPP
+
+#include "engine/algorithm.hpp"
+#include "engine/internal_route_result.hpp"
+#include "engine/phantom_node.hpp"
+#include "engine/routing_algorithms/alternative_path.hpp"
+#include "engine/routing_algorithms/direct_shortest_path.hpp"
+#include "engine/routing_algorithms/many_to_many.hpp"
+#include "engine/routing_algorithms/map_matching.hpp"
+#include "engine/routing_algorithms/shortest_path.hpp"
+#include "engine/routing_algorithms/tile_turns.hpp"
+
+namespace osrm
+{
+namespace engine
+{
+
+class RoutingAlgorithmsInterface
+{
+  public:
+    virtual InternalRouteResult
+    AlternativePathSearch(const PhantomNodes &phantom_node_pair) const = 0;
+
+    virtual InternalRouteResult
+    ShortestPathSearch(const std::vector<PhantomNodes> &phantom_node_pair,
+                       const boost::optional<bool> continue_straight_at_waypoint) const = 0;
+
+    virtual InternalRouteResult
+    DirectShortestPathSearch(const PhantomNodes &phantom_node_pair) const = 0;
+
+    virtual std::vector<EdgeWeight>
+    ManyToManySearch(const std::vector<PhantomNode> &phantom_nodes,
+                     const std::vector<std::size_t> &source_indices,
+                     const std::vector<std::size_t> &target_indices) const = 0;
+
+    virtual routing_algorithms::SubMatchingList
+    MapMatching(const routing_algorithms::CandidateLists &candidates_list,
+                const std::vector<util::Coordinate> &trace_coordinates,
+                const std::vector<unsigned> &trace_timestamps,
+                const std::vector<boost::optional<double>> &trace_gps_precision) const = 0;
+
+    virtual std::vector<routing_algorithms::TurnData>
+    GetTileTurns(const std::vector<datafacade::BaseDataFacade::RTreeLeaf> &edges,
+                 const std::vector<std::size_t> &sorted_edge_indexes) const = 0;
+
+    virtual bool HasAlternativePathSearch() const = 0;
+    virtual bool HasShortestPathSearch() const = 0;
+    virtual bool HasDirectShortestPathSearch() const = 0;
+    virtual bool HasMapMatching() const = 0;
+    virtual bool HasManyToManySearch() const = 0;
+    virtual bool HasGetTileTurns() const = 0;
+};
+
+namespace detail
+{
+
+struct NotImplementedException : public std::runtime_error
+{
+};
+}
+
+// Short-lived object passed to each plugin in request to wrap routing algorithms
+template <typename AlgorithmT> class RoutingAlgorithms final : public RoutingAlgorithmsInterface
+{
+  public:
+    RoutingAlgorithms(SearchEngineData &heaps,
+                      const datafacade::ContiguousInternalMemoryDataFacade<AlgorithmT> &facade)
+        : heaps(heaps), facade(facade)
+    {
+    }
+
+    virtual ~RoutingAlgorithms() = default;
```

Comment:
Yeah virtual destructor is required in base class for polymorphic delete - you can even make it

    virtual ~Base() = default;